### DISCUSSION TITLE
Passing Env variables/values for docker image via Admin UI

### DISCUSSION THREAD

**Discussion by tomg-formpak** [awslabs/aws-saas-boost#465] [Ideas]
@PoeppingT @brtrvn 
There are two use-case scenarios: 
1.	Add/Update Env variables common to the application/all tenants -  via `Application UI`
2.	Add/update Env variables specific for each tenant

- In 1st scenario, by default, the variables such as AWS_REGION, DB_HOST, etc are added to the ECS task definition via SaaS Boost, `Application UI`. But if the application needs to add extra required Env variables, we need to add that manually in the ECS task definition. 

**Suggestion**
The suggested option was to hardcode the KEY-VALUES pairs within the application. But similar to variables AWS_REGION, DB_HOST, DB_NAME, DB_PORT etc., Can we have an option (KEY-VALUE) in the `Application UI` in the dashboard to add user-defined environment variables required for the docker image, so that it’s added to the ECS task/SSM automatically during application deployment/update? 

Also once the application is deployed/updated and the tenant is onboard, to update/ add new variables later which is common to all tenants, how to pass those variables to the application (from SaaS Boost UI to the running application, so that it gets updated in SSM/ECS task definitions)? Do we need to add/update those manually in ECS tasks or SSM or both? Any suggestions for this? 

- In 2nd scenario, for example, passing tenant license or other tenant-specific env variables, we have to add this manually now. 

**Suggestion**
The suggestion was to use “upload config-file” option in the `Provision Tenant` tab. This suggested feature of the zip file option is good that can be used for other scenarios like adding CSS files, images or attachments in a zip file and the docker application can have the functionalities to read that from the tenant’s S3 bucket.  

But let’s say, to add env variables for a particular tenant, can we have an option of KEY – VALUE on the `Provision Tenant` tab in the `Onboarding page UI` so that it’s added to SSM/ECS tasks and later we can change/update those values from the `Tenant Detail` in the `Tenant UI` page for that particular tenant. 

Thank you

**Comment by brtrvn** [comment]
Thanks for the feedback @tomg-formpak. There is some of this mechanism today, but application-wide environment variables has not been on our radar. Can you enter a feature request for just that part?

As you say, you can upload a ZIP file as part of onboarding _each_ tenant and place in that archive whatever you like. For example, you could create a `*.properties` or `*.ini` file filled with tenant-specific settings to initialize your application with. As we've discussed in the past, your software executing in the docker container will have to download this ZIP file from the S3 bucket during initialization and apply its contents accordingly.

There is also (currently undocumented) an `attributes` property of all tenants. There is no UI for this feature right now in the Onboarding screen, but each tenant can have a list of arbitrary key/value pairs. If you were to utilize this, similar to the config ZIP file, your software would have to call the SaaS Boost tenant API and fetch the tenant details in order to access this list of key/value pairs.

We are considering [expanding on this](https://github.com/awslabs/aws-saas-boost/issues/323) `attributes` functionality to be able to mark some key/value pairs as "environment variables" and then automatically expose those to the ECS tasks. We're also considering whether one could mark some key/value pairs as "secrets" and have them encrypted at rest and stored in SecretsManager. There are still out standing design questions about any of this including the fact that because it's a tenant-level property all key/values/secrets would be exposed to _all_ services at runtime and that might not be what you want.
